File: TypeScript/124.二叉树中的最大路径和.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

let ans: number = Number.MIN_SAFE_INTEGER

function maxPathSum(root: TreeNode | null): number {
    ans = Number.MIN_SAFE_INTEGER
    maxPath(root)
    return ans
};

function maxPath(root: TreeNode | null): number {
    if (root === null)
        return 0
    const left = Math.max(0, maxPath(root.left))
    const right = Math.max(0, maxPath(root.right))
    const newPath = root.val + left + right
    ans = Math.max(ans, newPath)
    return root.val + Math.max(left, right)
}

File: Golang/124.二叉树中的最大路径和.go
package leetcode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	ans := -1 << 31
	maxPath(root, &ans)
	return ans
}

func maxPath(root *TreeNode, final_ans *int) int {
	if root == nil {
		return 0
	}
	left := maxInt(0, maxPath(root.Left, final_ans))
	right := maxInt(0, maxPath(root.Right, final_ans))
	newPath := root.Val + left + right
	*final_ans = maxInt(*final_ans, newPath)
	return root.Val + maxInt(left, right)
}

func maxInt(x int, y int) int {
	if x > y {
		return x
	}
	return y
}
